pkg/azure/priority_utils.go

```diff
@@ -17,40 +17,46 @@ limitations under the License.
 package azure
 
 import (
-	"strings"
+	"math"
 
 	"github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/network/armnetwork/v4"
 )
 
 // setupMaps fills the reservedPrioritiesInbound and reservedPrioritiesOutbound maps with the priorities of the existing rules in the NSG
 // This is done to avoid priorities conflicts when creating new rules
 // Existing rules map is filled to ensure that rules that just need their contents updated do not get recreated with new priorities
-func setupMaps(reservedPrioritiesInbound map[int32]bool, reservedPrioritiesOutbound map[int32]bool, existingRulePriorities map[string]int32, nsg *armnetwork.SecurityGroup) error {
+func setupMaps(reservedPrioritiesInbound map[int32]*armnetwork.SecurityRule, reservedPrioritiesOutbound map[int32]*armnetwork.SecurityRule, existingRulePriorities map[string]int32, nsg *armnetwork.SecurityGroup) error {
 	for _, rule := range nsg.Properties.SecurityRules {
 		if *rule.Properties.Direction == armnetwork.SecurityRuleDirectionInbound {
-			reservedPrioritiesInbound[*rule.Properties.Priority] = true
+			reservedPrioritiesInbound[*rule.Properties.Priority] = rule
 		} else if *rule.Properties.Direction == armnetwork.SecurityRuleDirectionOutbound {
-			reservedPrioritiesOutbound[*rule.Properties.Priority] = true
+			reservedPrioritiesOutbound[*rule.Properties.Priority] = rule
 		}
 
-		// skip rules that are not created by Paraglider, because some rules are added by default and have
-		// different fields such as port ranges which is not supported by Paraglider at the moment
-		if !strings.HasPrefix(*rule.Name, paragliderPrefix) {
+		// skip rules that are added by default, because they may have different fields
+		// such as port ranges which is not supported by Paraglider at the moment
+		if existingRulePriorities == nil || *rule.Properties.Priority > maxPriority {
 			continue
 		}
 		existingRulePriorities[*rule.Name] = *rule.Properties.Priority
 	}
 	return nil
 }
 
-// getPriority returns the next available priority that is not used by other rules
-func getPriority(reservedPriorities map[int32]bool, start int32, end int32) int32 {
-	var i int32
-	for i = start; i < end; i++ {
-		if !reservedPriorities[i] {
-			reservedPriorities[i] = true
+// getPriority returns the next available priority number that is not used by other rules
+func getPriority(reservedPriorities map[int32]*armnetwork.SecurityRule, start int32, end int32, ascendingSearch bool) int32 {
```

Comment:
Also, this interface is a little awkward I think. Maybe even just calling `ascendingSearch` something like `lowestAvailable` or something so that its less implementation-specific. And either way, I think you should add to the docstring with information about the function arguments.